kvgraph/kvgraph.go

```diff
@@ -85,52 +85,58 @@ func (kgdb *KVInterfaceGDB) GetTimestamp() string {
 // SetVertex adds an edge to the graph, if it already exists
 // in the graph, it is replaced
 func (kgdb *KVInterfaceGDB) SetVertex(vertexArray []*aql.Vertex) error {
-	for _, vertex := range vertexArray {
-		d, _ := proto.Marshal(vertex)
-		k := VertexKey(kgdb.graph, vertex.Gid)
-		err := kgdb.kv.Set(k, d)
-		if err != nil {
-			return err
+	kgdb.kv.Update(func(tx KVTransaction) error {
+		for _, vertex := range vertexArray {
+			d, _ := proto.Marshal(vertex)
+			k := VertexKey(kgdb.graph, vertex.Gid)
+			err := tx.Set(k, d)
+			if err != nil {
+				return err
+			}
+			kgdb.ts.Touch(kgdb.graph)
 		}
-		kgdb.ts.Touch(kgdb.graph)
-	}
+		return nil
+	})
 	return nil
 }
 
 // SetEdge adds an edge to the graph, if the id is not "" and in already exists
 // in the graph, it is replaced
 func (kgdb *KVInterfaceGDB) SetEdge(edgeArray []*aql.Edge) error {
-	for _, edge := range edgeArray {
-		if edge.Gid == "" {
-			eid := fmt.Sprintf("%d", rand.Uint64())
-			for ; kgdb.kv.HasKey(EdgeKeyPrefix(kgdb.graph, eid)); eid = fmt.Sprintf("%d", rand.Uint64()) {
+	kgdb.kv.Update(func(tx KVTransaction) error {
+		for _, edge := range edgeArray {
+			if edge.Gid == "" {
+				eid := fmt.Sprintf("%d", rand.Uint64())
+				for ; tx.HasKey(EdgeKeyPrefix(kgdb.graph, eid)); eid = fmt.Sprintf("%d", rand.Uint64()) {
```

Comment:
What's going on here? This is cryptic.

Reply:
Random key assignment

Reply:
A helper function called "randomKeyAssignment" would be easy.